Enable updates for labels and annotations

jonnylangefeld opened the pull request:
Fix #1386

jonnylangefeld commented on pkg/router/util.go:
```diff
@@ -32,3 +32,18 @@ func filterMetadata(meta map[string]string) map[string]string {
 }
+
+// mergeMeta merges two maps with the second one overwriting the first one
+func mergeMeta(m1 map[string]string, m2 map[string]string) map[string]string {
```
If anyone knows a more Kubernetes native function that already exists I'm happy to use that. I just couldn't find any (even though I thought I used one in the past)

jonnylangefeld commented:
aparently keda 2.10.1, released on 3/14 only allows for Kubernetes `>=v1.24.0-0` while keda 2.10.0, released on 3/9 allows for Kubernetes `>=v1.23.0-0`, which is what is used in integration tests. For now I pinned the keda version until the integration test k8s will get updated.

jonnylangefeld commented:
@stefanprodan would you mind taking a look to get this merged?

aryan9600 reviewed (commented):
hello @jonnylangefeld, thanks for opening this PR! this issue is certainly valid, but i don't think the proposed changes fully fix it.
since we only update the `VirtualService` on the basis of `.spec` and not `.metadata`, the new labels and/or annotations will only get applied when the there's a difference in the spec of the `VirtualService`.
we should instead compare the metadata before proceeding to update it. asking @stefanprodan to also weigh in here.

jonnylangefeld commented:
@aryan9600 thanks for your review! You are right, the update only happened if there is a diff in the spec, not in the metadata. I addressed the issue in my last commit 50265e847d526d0ef4bfc5501944e9b15aaa3598 by adding a `diffMeta()` to the diff condition. That function checks if there's a diff in labels or annotations, but only for labels or annotations set by the apex, as there could be other labels and annotations on the resource that are coming from outside the controller.

Please take another look.

aryan9600 reviewed (changes requested):
please add these changes to `pkg/router/kuma.go` and `pkg/router/traefik.go` as well, thanks.

jonnylangefeld commented on test/keda/install.sh:
```diff
@@ -9,7 +9,7 @@ mkdir -p ${REPO_ROOT}/bin
 helm repo add kedacore https://kedacore.github.io/charts
 kubectl create ns keda
-helm install keda kedacore/keda --namespace keda --wait
+helm install keda kedacore/keda --namespace keda --wait --version 2.10.0
```
Hmm is keda not needed anymore? If we remove this line, then we should probably also remove the other lines to add the keda helm repo and the namespace creation.

This PR is obviously unrelated to keda, I just happened to run into a CI issue: https://github.com/fluxcd/flagger/pull/1392#issuecomment-1473929857

aryan9600 replied:
the issue has been fixed in #1406, hence this change is not required. please rebase your changes on top of main.

aryan9600 replied:
also sorry for the wrong suggestion, this is the correct one:

```suggestion
helm install keda kedacore/keda --namespace keda --wait
```

basically no changes should be made to this script, due to the reason stated above.

jonnylangefeld commented:
Thanks for your review @aryan9600 I addressed all you comments but am unsure how to handle the `keda` one. I wouldn't have touched that line if there wasn't a CI error without pinning the version. So maybe this can be addressed in another PR?

aryan9600 commented:
hey @jonnylangefeld, would it be possible to make the requested changes and get the PR in shape to be merged (needs squashing and rebasing) by today? that way we could include it in the upcoming release, otherwise this will be included in the release after that.

jonnylangefeld commented:
@aryan9600 that would be awesome if it could be included in the upcoming release. I resolved the last remaining comment and squashed and rebased with main.

aryan9600 reviewed (changes requested):
i was testing these changes and i found an issue. since we are merging the object metadata and the metadata specified in the Canary spec, deleting a label/annotation from the Canary spec metadata has no effect. steps to repro:

* apply a canary which includes the below bit:
```
  service:
    apex:
      annotations:
        test: annotation-test
        another: annotation
```
* wait for the canary to be reconciled and the annotations to appear on the `VirtualService`.
* delete `another: annotation` from `spec.service.apex.metadata.annotations` in the Canary.
* `kubectl describe virtualservice <vs-name>` will show that `another: annotation` still exists as annotation.

instead of merging the metadata, we should simply be overwriting the object's metadata with what's specified in the Canary. any edits to the it's metadata needs to be made through the Canary, since it owns the object.

jonnylangefeld commented:
@aryan9600 good point about that annotations and labels won't get deleted from the `virtualservice`. As you said, the reason for this is that we merge the metadata, rather than overwriting it.
The issue with overwriting is that it's going to impact other controllers that operate on the `virtualservice` resource. They might write a label/annotation to the `virtualservice` and if flagger keeps removing that, we'll end up in a controller conflict. We have one use case of this in-house, but I'm sure there are other use cases that we don't even know of as well. I think the right way to do this would be to use server-side-apply, but that would take a larger refactoring across the repo.

So this creates a dilemma and gives us a decision to make:
1. We could change flagger to own and always overwrite all the labels and annotations of the `virtualservice`, which will impact other controllers in the ecosystem that rely on writing labels/annotations to a `virtualservice`, potentially creating endless reconcile cycles.
2. We could accept that the Canary can only add or modify existing labels/annotations and cannot remove existing keys but don't have any risk of introducing infinite reconcile loops.

Given these considerations, would you still say we should change flagger to own all labels/annotations and removing all from other controllers and go with option 1? My preference would be option 2 and document the limitation but know that we don't run into endless loops.

Let me know what you think.

jonnylangefeld commented:
@aryan9600 if you are okay with option 2 above, I pushed a commit to update the limitation in the documentation.

aryan9600 commented:
@jonnylangefeld i'd lean towards option one. My reasoning is that since Flagger creates the object and manages it, it doesn't expect other controllers to operate on it. The _proper_ way to fix this is indeed server-side apply, but unfortunately you're correct about it being a huge refactor.
If a controller wants to modify the object's metadata it should do it by modifying the Canary itself. The Canary to be modified can be found by looking at the object's `.metadata.ownerRefs`. We can add this info to the FAQ page in the docs, if it helps.

@stefanprodan wdyt?

jonnylangefeld commented:
Okay, in the latest commit 13ed85a4518428445130b174148f584e937892cd we now always write the metadata from `canary.Spec.Service.Apex` to the `virtualservice`.

aryan9600 commented on pkg/router/appmesh.go:
```diff
@@ -187,9 +188,11 @@ func (ar *AppMeshRouter) reconcileVirtualNode(canary *flaggerv1.Canary, name str
 	// update virtual node
 	if virtualnode != nil {
-		if diff := cmp.Diff(vnSpec, virtualnode.Spec); diff != "" {
+		if diff := cmp.Diff(vnSpec, virtualnode.Spec); diff != "" || compareObjMeta(newMetadata, virtualnode) {
```
sorry i didn't catch this earlier, but i think it'd be a better idea to use `cmp.Diff` instead of our own util func, as it'd increase maintenance burden and we are already using it everywhere else to compare stuff.
so we can have something like :
```suggestion
		specDiff := cmp.Diff(vnSpec, virtualnode.Spec)
		labelsDiff := cmp.Diff(newMetadata.Labels, virtualNode.Metadata.Labels)
		annotationsDiff := cmp.Diff(newMetadata.Annotations, virtualNode.Metadata.Annotations)
	     if specDiff != "" || labelsDiff != "" || annotationsDiff != ""{
```
